Finished translation of State as a Snapshot (#549)
reactjs/uk.react.dev: src/content/learn/state-as-a-snapshot.md

---

> title: Стан як зняток

Author: Відбиток? Кадр?

> title: Стан як зняток

Reviewer: Нє, я би тут залишила снепшот або копія даних

> title: Стан як зняток

Reviewer: Але backup - це (резервна) копія даних, тому поки що снепшот (і додай заодно до словника у TRANSLATION).

снепшотів не багато серед файлів, так що потім за потреби змінимо

> title: Стан як зняток

Reviewer: ```suggestion
title: Стан як снепшот
```

> title: Стан як зняток

Author: Мені не подобається, тому що в такому вигляді ця назва нічого не говорить. Snapshot – це словникове слово англійської мови https://www.dictionary.com/browse/snapshot , яке саме в собі пояснює, яку природу має стан в React. А в українській цей трансліт не повідомляє геть нічого. Можливо, краще таки "кадр"?

> title: Стан як зняток

Reviewer: комп'ютер теж не говорив, але ж запозичили

кадр, по-моєму, французьке слово, не наше і запозичене, але розуміння прийшло згодом

> title: Стан як зняток

Reviewer: мені подобається кадр, бо є розкадрування і це щось схоже - окремий кадр із усього процесу. але можна використовувати кадр для пояснення, та не для технічної дефініції

> title: Стан як зняток

Author: То давай тоді лишимо "снепшот" усюди по тексту, але в поясненні пояснимо через кадр

> title: Стан як зняток

Reviewer: давай

https://github.com/reactjs/uk.react.dev/pull/549#discussion_r1690071310 запозичую не особисто я, вже й інші, не маючи відповідника зрозумілого, використовували снепшот і снапшот

> title: Стан як зняток

Reviewer: якби ми були перші, то я би снепшот не вводила

---

<Intro>

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Author: ```suggestion
Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як снепшот — миттєвий знімок. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.
```
Можливо, хоча б так можна було б надати "снепшоту" якийсь зміст

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Reviewer: тобто тут знімок ок, так?

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Author: Знімок, зняток, зліпок – варіанти, з яких мені до смаку припав саме зняток

Я розумію, що я цими додатковими варіантами наводжу тільки більший безлад, але я це роблю суто через те, що мені смертельно не подобається саме використання тут слова "копія", взагалі, й готовий пристати на різні варіанти

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Reviewer: тут на кадр

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Reviewer: чому один-єдиний, а не окремий?

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Author: ```suggestion
Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як снепшот – окремий незмінний кадр. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.
```
Нехай так

> Змінні стану можуть нагадувати звичайні змінні JavaScript, які можна зчитати та змінити. Проте стан поводиться радше як зняток. Задання йому значення не змінює наявну змінну стану, а натомість запускає повторний рендер.

Reviewer: ок

</Intro>

<YouWillLearn>

* Як задання стану запускає повторні рендери
* Коли та як оновлюється стан

> * Чому стан не оновлюється зразу після його задання

Reviewer: зразу - калька
```suggestion
* Чому стан не оновлюється відразу після задання значення
```

</Sandpack>

Ось що відбувається, коли ви клацаєте кнопку:

1. Виконується обробник події `onSubmit`.
2. `setIsSent(true)` задає `isSent` значення `true`, і таким чином додає в чергу новий рендер.
3. React виконує повторний рендер компонента, згідно з новим значенням `isSent`.

Погляньмо уважніше на взаємозв'язок між станом і рендерингом.

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Author: ```suggestion
## Рендеринг бере один снепшот за раз {/*rendering-takes-a-snapshot-in-time*/}
```

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Reviewer: чому ти перекладаєш takes тут як бере?

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Author: Бо тут треба перекласти речення, яке означає "В межах одного рендеру діє лише один снепшот-кадр-зняток-знімок стану". Що рендер робить зі снепшотом стану? Здається, бере

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Reviewer: takes це ще потребує, потрібно, забрати, приймати, знадобилось, вимірювати

Рендеринг бере один снепшот за раз - це дуже розмовно і виглядає дивно

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Reviewer: Ще Takes snapshot - сapture snapshot - створює снепшот/кадр/знімок або "зняв" або "взяв"

```suggestion
## Кожен рендер отримує новий снепшот {/*rendering-takes-a-snapshot-in-time*/}
```
aбо
```suggestion
## Кожен рендер створює новий снепшот {/*rendering-takes-a-snapshot-in-time*/}
```
або
```suggestion
## Для рендерингу потрібен новий снепшот {/*rendering-takes-a-snapshot-in-time*/}
```

> ## Рендеринг бере один зняток за раз {/*rendering-takes-a-snapshot-in-time*/}

Reviewer: ChatGPT каже
Рендеринг фіксує знімок у часі
цікаво

> ["Рендеринг"](/learn/render-and-commit#step-2-react-renders-your-components) означає, що React викликає ваш компонент, який є функцією. JSX, який ви повертаєте з цієї функції, — це як зняток UI в певну мить часу. Його пропси, обробники подій і локальні змінні обчислюються **з використанням його стану в мить рендеру.**

Author: По-моєму, копія без оригіналу це не дуже копія

> ["Рендеринг"](/learn/render-and-commit#step-2-react-renders-your-components) означає, що React викликає ваш компонент, який є функцією. JSX, який ви повертаєте з цієї функції, — це як зняток UI в певну мить часу. Його пропси, обробники подій і локальні змінні обчислюються **з використанням його стану в мить рендеру.**

Reviewer: отут згодна на кадр

> ["Рендеринг"](/learn/render-and-commit#step-2-react-renders-your-components) означає, що React викликає ваш компонент, який є функцією. JSX, який ви повертаєте з цієї функції, — це як зняток UI в певну мить часу. Його пропси, обробники подій і локальні змінні обчислюються **з використанням його стану в мить рендеру.**

Author: ```suggestion
["Рендеринг"](/learn/render-and-commit#step-2-react-renders-your-components) означає, що React викликає ваш компонент, який є функцією. JSX, який ви повертаєте з цієї функції, — це як кадр UI в певну мить часу. Його пропси, обробники подій і локальні змінні обчислюються **з використанням його стану в мить рендеру.**
```

}, 3000);
```

Стан, що зберігається в React, може змінитися на час виведення повідомлення, але він був запланований з використанням знятку стану, актуального тоді, коли користувач повзаємодіяв з елементом!

**Значення змінної стану ніколи не змінюється протягом одного рендеру,** навіть якщо код обробника події є асинхронним. Всередині `onClick` *поточного рендеру* значення `number` усе одно залишається `0`, навіть після виклику `setNumber(number + 5)`. Її значення "зафіксувалося", коли React "зробив зняток" UI, викликавши ваш компонент.

Ось приклад того, що робить обробники помилок більш захищеними щодо помилок хронометражу. Нижче — форма, що надсилає повідомлення з п'ятисекундною затримкою. Уявіть такий сценарій:

1. Ви натискаєте кнопку "Надіслати", надсилаючи "Привіт" Анні.
2. Перш ніж закінчиться п'ятисекундна затримка, ви змінюєте значення в полі "Кому" на "Богдан".

Як гадаєте, що покаже `alert`? Чи виведеться "Ви надіслали Привіт користувачу Анна"? Чи, можливо, "Ви надіслали Привіт користувачу Богдан"? Спробуйте вгадати на основі того, що знаєте, а потім перевірте:

<Sandpack>

```js
import { useState } from 'react';

export default function Form() {

>   const [to, setTo] = useState('Анна');

Reviewer: Аліса і Боб (якраз А і Б)

>   const [to, setTo] = useState('Анна');

Reviewer: точно не Анна (бо Ганна)

>   const [to, setTo] = useState('Анна');

Reviewer: а тут шо робимо?

>   const [to, setTo] = useState('Анна');

Reviewer: ага, ти змінив, закриваю

</Sandpack>

**React підтримує значення стану "зафіксованими" в межах обробників подій одного рендеру.** Немає потреби перейматися тим, чи стан змінився, поки виконувався код.

Але що якщо хочеться зчитати найсвіжіший стан перед повторним рендером? Знадобиться скористатися [функцією-оновлювачем стану](/learn/queueing-a-series-of-state-updates), про яку мова піде на наступній сторінці!

<Recap>

* Задання стану спричиняє прохання про новий рендер.
* React зберігає стан поза вашим компонентом, неначе на поличці.
* Коли викликається `useState`, React видає зняток стану *для конкретного поточного рендеру*.
* Змінні та обробники подій не "переживають" повторні рендери. Кожний рендер має власні обробники подій.
* Кожний рендер (а також функції в ньому) завжди "бачить" зняток стану, який React видав *цьому конкретному* рендеру.
* Можна уявно підставити стан у обробниках подій, подібного до того, як ми уявляємо JSX після рендеру.
* Обробники подій, створені в минулому, мають значення стану з тих рендерів, у яких вони створені.

</Recap>

> 

Author: ```suggestion

```

* React зберігає стан поза вашим компонентом, неначе на поличці.
* Коли викликається `useState`, React видає зняток стану *для конкретного поточного рендеру*.
* Змінні та обробники подій не "переживають" повторні рендери. Кожний рендер має власні обробники подій.
* Кожний рендер (а також функції в ньому) завжди "бачить" зняток стану, який React видав *цьому конкретному* рендеру.
* Можна уявно підставити стан у обробниках подій, подібного до того, як ми уявляємо JSX після рендеру.
* Обробники подій, створені в минулому, мають значення стану з тих рендерів, у яких вони створені.

</Recap>

> 

Reviewer: Same about line-by-line parity https://github.com/reactjs/uk.react.dev/pull/547#discussion_r1677127928
